include/boost/serialization/singleton.hpp

```diff
@@ -91,75 +98,102 @@ class BOOST_SYMBOL_VISIBLE singleton_module :
+    BOOST_DLLEXPORT static bool & get_lock() BOOST_USED {
         static bool lock = false;
         return lock;
     }
 
 public:
-    BOOST_DLLEXPORT void lock(){
+    BOOST_DLLEXPORT static void lock(){
         get_lock() = true;
     }
-    BOOST_DLLEXPORT void unlock(){
+    BOOST_DLLEXPORT static void unlock(){
         get_lock() = false;
     }
-    BOOST_DLLEXPORT bool is_locked(){
+    BOOST_DLLEXPORT static bool is_locked(){
         return get_lock();
     }
 };
 
-static inline singleton_module & get_singleton_module(){
-    static singleton_module m;
-    return m;
-}
+namespace detail {
+
+// This is the class actually instantiated and hence the real singleton.
+// So there will only be one instance of this class. This does not hold
+// for singleton<T> as a class derived from singleton<T> could be
+// instantiated multiple times.
+// It also provides a flag `is_destroyed` which returns true, when the
+// class was destructed. It is static and hence accesible even after
+// destruction. This can be used to check, if the singleton is still
+// accesible e.g. in destructors of other singletons.
+template<class T>
+class singleton_wrapper : public T
+{
+    static bool & get_is_destroyed(){
+        // Prefer a static function member to avoid LNK1179.
+        // Note: As this is for a singleton (1 instance only) it must be set
+        // never be reset (to false)!
+        static bool is_destroyed_flag = false;
+        return is_destroyed_flag;
+    }
+public:
+    singleton_wrapper(){
+        BOOST_ASSERT(!is_destroyed());
+    }
+    ~singleton_wrapper(){
+        get_is_destroyed() = true;
+    }
+    static bool is_destroyed(){
+        return get_is_destroyed();
+    }
+};
+
+} // detail
 
 template <class T>
-class singleton {
+class singleton : public singleton_module
+{
 private:
-    // note presumption that T has a default constructor
     static T & m_instance;
     // include this to provoke instantiation at pre-execution time
-    static void use(T const &) {}
+    static void use(T const *) {}
     static T & get_instance() {
-        static T t;
+        BOOST_ASSERT(!is_destroyed());
 
-        // refer to instance, causing it to be instantiated (and
-        // initialized at startup on working compilers)
-        BOOST_ASSERT(! is_destroyed());
+        // use a wrapper so that types T with protected constructors can be used
+        // Using a static function member avoids LNK1179
+        static detail::singleton_wrapper< T > t;
 
         // note that the following is absolutely essential.
         // commenting out this statement will cause compilers to fail to
         // construct the instance at pre-execution time.  This would prevent
         // our usage/implementation of "locking" and introduce uncertainty into
-        // the sequence of object initializaition.
-        use(m_instance);
+        // the sequence of object initialization.
+        use(& m_instance);
```

Comment:
`use(m_instance)` - as before this patch - is undefined behavior, but I think that this one is undefined too. Non-undefined would have been to use a pointer as m_instance. Of course getting this past Robert may prove problematic. :-)

Reply:
Can you explain why it is UB?

I do think this line is not required because of https://github.com/boostorg/serialization/pull/105/files/53392794c49ba129c4f3eb70debb7426746b325b#diff-3f1e861be15fd4926f5e5b00388f2309R198 but I'm to afraid to remove it ;)

Reply:
In either case, I don't see how this is going to work - `use` will just get inlined here and disappear. But who knows.

Reply:
At the point the initializer for m_instance is called, m_instance is not yet initialized, and using a reference that does not refer to a valid object is undefined behavior.

A pointer would not have this issue because globals are zero-initialized first, before the dynamic initializers are run, so the value of m_instance when accessed in its own initializer would be nullptr.

Even better would probably be to use a pointer for m_instance and `use(&m_instance)`, which avoids all these subtleties, as the address of m_instance is always valid no matter whether it's initialized or not.

Reply:
Wait. My patch uses `&m_instance`, hence it takes the address of the object. This is never dereferenced. With your statement 

>  as the address of m_instance is always valid no matter whether it's initialized or not.

This is ok, isn't it?

So at no point a reference is used, we only get the address of a static member.

Reply:
`m_instance` is an uninitialized reference. `&m_instance` is undefined in that case.

Reply:
Actually I don't understand WHY it is required. Just judging from the comment this is to "construct the instance at pre-execution time". But the `m_instance` with its initialization is there for exactly the same reason (I added this as a comment with a link to https://groups.google.com/forum/#!topic/microsoft.public.vc.language/kDVNLnIsfZk) 

But @robertramey wrote in an email:

> LOL - it's only in there because the tests fail without it.

I verified this locally (Linux, GCC), but don't understand it...

I'm currently testing `static T * m_instance;` and it seems to work. But this would be a separate PR.